File: android/app/src/main/kotlin/com/example/panel_app/Amf3.kt
package com.example.panel_app

import java.io.ByteArrayOutputStream
import java.nio.charset.Charset
import java.util.Date

/**
 * Minimal AMF3 encoder/decoder for the subset of types
 * used by the Ninja Sage workflow (String, num, bool,
 * List, Map, Date, ByteArray).
 *
 * This is intentionally self-contained and does not depend
 * on any external AMF libraries.
 */
object Amf3 {
    const val UNDEFINED_TYPE = 0
    const val NULL_TYPE = 1
    const val FALSE_TYPE = 2
    const val TRUE_TYPE = 3
    const val INTEGER_TYPE = 4
    const val DOUBLE_TYPE = 5
    const val STRING_TYPE = 6
    const val DATE_TYPE = 8
    const val ARRAY_TYPE = 9
    const val OBJECT_TYPE = 10
    const val BYTEARRAY_TYPE = 12
    const val AMF0_AMF3 = 17

    private const val UINT29_MASK = 0x1FFFFFFF
    private const val INT28_MAX_VALUE = 0x0FFFFFFF
    private const val INT28_MIN_VALUE = -0x0FFFFFFF

    class Writer {
        private val buffer = ByteArrayOutputStream()

        fun toByteArray(): ByteArray = buffer.toByteArray()

        fun writeByte(v: Int) {
            buffer.write(v and 0xFF)
        }

        fun writeShort(v: Int) {
            writeByte((v ushr 8) and 0xFF)
            writeByte(v and 0xFF)
        }

        fun writeInt(v: Int) {
            writeByte((v ushr 24) and 0xFF)
            writeByte((v ushr 16) and 0xFF)
            writeByte((v ushr 8) and 0xFF)
            writeByte(v and 0xFF)
        }

        fun writeUInt29(value: Int) {
            var v = value and UINT29_MASK
            when {
                v < 0x80 -> writeByte(v)
                v < 0x4000 -> {
                    writeByte(((v shr 7) and 0x7F) or 0x80)
                    writeByte(v and 0x7F)
                }
                v < 0x200000 -> {
                    writeByte(((v shr 14) and 0x7F) or 0x80)
                    writeByte(((v shr 7) and 0x7F) or 0x80)
                    writeByte(v and 0x7F)
                }
                v < 0x40000000 -> {
                    writeByte(((v shr 22) and 0x7F) or 0x80)
                    writeByte(((v shr 15) and 0x7F) or 0x80)
                    writeByte(((v shr 8) and 0x7F) or 0x80)
                    writeByte(v and 0xFF)
                }
                else -> error("Integer out of range: $value")
            }
        }

        fun writeBytes(bytes: ByteArray) {
            buffer.write(bytes)
        }

        fun writeUTF(str: String, asAmf: Boolean) {
            val bytes = str.toByteArray(Charset.forName("UTF-8"))
            if (asAmf) {
                writeUInt29((bytes.size shl 1) or 1)
                writeBytes(bytes)
            } else {
                writeShort(bytes.size)
                writeBytes(bytes)
            }
        }

        fun writeStringWithoutType(str: String) {
            if (str.isEmpty()) {
                // empty inline string
                writeUInt29(1)
            } else {
                // always inline; no reference table
                writeUTF(str, true)
            }
        }

        private fun writeAmfInt(v: Int) {
            if (v in INT28_MIN_VALUE..INT28_MAX_VALUE) {
                val masked = v and UINT29_MASK
                writeByte(INTEGER_TYPE)
                writeUInt29(masked)
            } else {
                writeByte(DOUBLE_TYPE)
                writeDouble(v.toDouble())
            }
        }

        private fun writeDouble(value: Double) {
            val bits = java.lang.Double.doubleToRawLongBits(value)
            for (i in 7 downTo 0) {
                writeByte(((bits shr (8 * i)) and 0xFF).toInt())
            }
        }

        fun writeObject(value: Any?) {
            when (value) {
                null -> writeByte(NULL_TYPE)
                is String -> {
                    writeByte(STRING_TYPE)
                    writeStringWithoutType(value)
                }
                is Boolean -> writeByte(if (value) TRUE_TYPE else FALSE_TYPE)
                is Int, is Short, is Byte -> writeAmfInt((value as Number).toInt())
                is Long -> {
                    val v = value.toInt()
                    writeAmfInt(v)
                }
                is Number -> {
                    writeByte(DOUBLE_TYPE)
                    writeDouble(value.toDouble())
                }
                is List<*> -> writeArray(value)
                is Map<*, *> -> writeMap(value)
                is Date -> {
                    writeByte(DATE_TYPE)
                    writeUInt29(1) // inline
                    writeDouble(value.time.toDouble())
                }
                is ByteArray -> {
                    writeByte(BYTEARRAY_TYPE)
                    writeUInt29((value.size shl 1) or 1)
                    writeBytes(value)
                }
                else -> {
                    // Fallback: encode as string
                    val s = value.toString()
                    writeByte(STRING_TYPE)
                    writeStringWithoutType(s)
                }
            }
        }

        private fun writeArray(list: List<*>?) {
            writeByte(ARRAY_TYPE)
            val items = list ?: emptyList<Any?>()
            writeUInt29((items.size shl 1) or 1) // inline, length
            writeUInt29(1) // empty string -> no named keys
            for (item in items) {
                writeObject(item)
            }
        }

        @Suppress("UNCHECKED_CAST")
        private fun writeMap(map: Map<*, *>?) {
            writeByte(OBJECT_TYPE)
            val entries = map ?: emptyMap<Any?, Any?>()
            // Inline object, dynamic, zero sealed properties.
            writeUInt29(0x0B)
            writeStringWithoutType("") // class name
            for ((k, v) in entries) {
                val key = k?.toString() ?: ""
                writeStringWithoutType(key)
                writeObject(v)
            }
            writeStringWithoutType("") // end of dynamic members
        }
    }

    private class Reader(private val data: ByteArray) {
        private var pos: Int = 0
        private val objects = mutableListOf<Any?>()
        private val strings = mutableListOf<String>()
        private val traits = mutableListOf<Traits>()

        data class Traits(
            val className: String?,
            val externalizable: Boolean,
            val dynamic: Boolean,
            val props: List<String>,
        )

        fun read(): Int {
            return data[pos++].toInt() and 0xFF
        }

        fun readUnsignedShort(): Int {
            val c1 = read()
            val c2 = read()
            return (c1 shl 8) or c2
        }

        fun readUInt29(): Int {
            var b = read()
            if (b < 0x80) return b
            var value = (b and 0x7F) shl 7
            b = read()
            if (b < 0x80) return value or b
            value = (value or (b and 0x7F)) shl 7
            b = read()
            if (b < 0x80) return value or b
            value = (value or (b and 0x7F)) shl 8
            b = read()
            return value or b
        }

        private fun readBytes(length: Int): ByteArray {
            val result = data.copyOfRange(pos, pos + length)
            pos += length
            return result
        }

        fun readUTF(length: Int? = null): String {
            val len = length ?: readUnsignedShort()
            if (len == 0) return ""
            val bytes = readBytes(len)
            return String(bytes, Charset.forName("UTF-8"))
        }

        fun reset() {
            objects.clear()
            strings.clear()
            traits.clear()
        }

        fun readObject(): Any? {
            val type = read()
            return readObjectValue(type)
        }

        fun readHeaderObject(): Any? {
            var type = read()
            if (type == AMF0_AMF3) {
                type = read()
                return readObjectValue(type)
            }
            return readHeaderObjectValue(type)
        }

        private fun readString(): String {
            val ref = readUInt29()
            if (ref and 1 == 0) {
                val index = ref shr 1
                return strings[index]
            }
            val len = ref shr 1
            if (len == 0) return ""
            val str = readUTF(len)
            strings.add(str)
            return str
        }

        private fun rememberObject(value: Any?) {
            objects.add(value)
        }

        private fun getObject(index: Int): Any? {
            return objects[index]
        }

        private fun rememberTraits(t: Traits) {
            traits.add(t)
        }

        private fun getTraits(index: Int): Traits {
            return traits[index]
        }

        private fun readTraits(ref: Int): Traits {
            return if (ref and 3 == 1) {
                getTraits(ref shr 2)
            } else {
                val count = ref shr 4
                val className = readString().takeIf { it.isNotEmpty() }
                val externalizable = (ref and 4) == 4
                val dynamic = (ref and 8) == 8
                val props = mutableListOf<String>()
                repeat(count) {
                    props.add(readString())
                }
                val traits = Traits(
                    className = className,
                    externalizable = externalizable,
                    dynamic = dynamic,
                    props = props,
                )
                rememberTraits(traits)
                traits
            }
        }

        private fun readScriptObject(): Any? {
            val ref = readUInt29()
            if (ref and 1 == 0) {
                return getObject(ref shr 1)
            }

            val traits = readTraits(ref)
            val map = mutableMapOf<String, Any?>()
            rememberObject(map)

            if (traits.externalizable) {
                // limited support: handle ArrayCollection and simple maps if needed
                // For game payloads we do not expect externalizable objects,
                // so this path is intentionally minimal.
                return map
            }

            for (prop in traits.props) {
                val value = readObject()
                map[prop] = value
            }

            if (traits.dynamic) {
                while (true) {
                    val name = readString()
                    if (name.isEmpty()) break
                    map[name] = readObject()
                }
            }

            return map
        }

        private fun readArray(): Any? {
            val ref = readUInt29()
            if (ref and 1 == 0) {
                return getObject(ref shr 1)
            }
            val len = ref shr 1

            var map: MutableMap<String, Any?>? = null
            while (true) {
                val name = readString()
                if (name.isEmpty()) break
                if (map == null) {
                    map = mutableMapOf()
                    rememberObject(map)
                }
                map[name] = readObject()
            }

            if (map == null) {
                val list = MutableList<Any?>(len) { null }
                rememberObject(list)
                for (i in 0 until len) {
                    list[i] = readObject()
                }
                return list
            } else {
                for (i in 0 until len) {
                    map[i.toString()] = readObject()
                }
                return map
            }
        }

        private fun readDouble(): Double {
            val b0 = read()
            val b1 = read()
            val b2 = read()
            val b3 = read()
            val b4 = read()
            val b5 = read()
            val b6 = read()
            val b7 = read()
            val bits =
                (b0.toLong() shl 56) or
                    (b1.toLong() shl 48) or
                    (b2.toLong() shl 40) or
                    (b3.toLong() shl 32) or
                    (b4.toLong() shl 24) or
                    (b5.toLong() shl 16) or
                    (b6.toLong() shl 8) or
                    b7.toLong()
            return java.lang.Double.longBitsToDouble(bits)
        }

        private fun readDate(): Any? {
            val ref = readUInt29()
            if (ref and 1 == 0) {
                return getObject(ref shr 1)
            }
            val time = readDouble().toLong()
            val date = Date(time)
            rememberObject(date)
            return date
        }

        private fun readMap(): Map<String, Any?> {
            val ref = readUInt29()
            if (ref and 1 == 0) {
                @Suppress("UNCHECKED_CAST")
                return getObject(ref shr 1) as Map<String, Any?>
            }
            val length = ref shr 1
            val map = mutableMapOf<String, Any?>()
            if (length > 0) {
                rememberObject(map)
                while (true) {
                    val name = readObject() as? String ?: break
                    map[name] = readObject()
                }
            }
            return map
        }

        private fun readByteArray(): ByteArray {
            val ref = readUInt29()
            if (ref and 1 == 0) {
                @Suppress("UNCHECKED_CAST")
                return getObject(ref shr 1) as ByteArray
            }
            val len = ref shr 1
            val ba = readBytes(len)
            rememberObject(ba)
            return ba
        }

        private fun readHeaderObjectValue(type: Int): Any? {
            return when (type) {
                2 -> readUTF()
                else -> error("Unknown AMF0 header type: $type")
            }
        }

        private fun readObjectValue(type: Int): Any? {
            return when (type) {
                STRING_TYPE -> readString()
                OBJECT_TYPE -> readScriptObject()
                ARRAY_TYPE -> readArray()
                FALSE_TYPE -> false
                TRUE_TYPE -> true
                INTEGER_TYPE -> {
                    val tmp = readUInt29()
                    (tmp shl 3) shr 3
                }
                DOUBLE_TYPE -> readDouble()
                UNDEFINED_TYPE, NULL_TYPE -> null
                DATE_TYPE -> readDate()
                BYTEARRAY_TYPE -> readByteArray()
                AMF0_AMF3 -> readObject()
                else -> error("Unknown AMF3 type: $type")
            }
        }
    }

    data class MessageBody(
        val targetURI: String,
        val responseURI: String,
        val data: Any?,
    )

    private class Deserializer(private val reader: Reader) {
        fun readFirstBody(): MessageBody? {
            reader.readUnsignedShort() // version

            val headerCount = reader.readUnsignedShort()
            repeat(headerCount) {
                readHeader()
            }

            val bodyCount = reader.readUnsignedShort()
            if (bodyCount == 0) return null
            return readBody()
        }

        private fun readHeader() {
            reader.readUTF() // name
            reader.read() // mustUnderstand
            repeat(4) { reader.read() } // length
            reader.reset()
            reader.readHeaderObject()
        }

        private fun readBody(): MessageBody {
            val target = reader.readUTF()
            val response = reader.readUTF()
            repeat(4) { reader.read() } // length
            reader.reset()
            val data = reader.readObject()
            return MessageBody(target, response, data)
        }
    }

    fun decodeFirstBody(data: ByteArray): MessageBody? {
        val reader = Reader(data)
        val deserializer = Deserializer(reader)
        return deserializer.readFirstBody()
    }

    data class ActionMessage(
        val version: Int = 3,
        val bodies: List<MessageBody>,
    )

    class Serializer {
        fun writeMessage(message: ActionMessage): ByteArray {
            val writer = Writer()
            writer.writeShort(message.version)
            writer.writeShort(0) // header count
            writer.writeShort(message.bodies.size)

            for (body in message.bodies) {
                writeBody(writer, body)
            }

            return writer.toByteArray()
        }

        private fun writeBody(writer: Writer, body: MessageBody) {
            writer.writeUTF(body.targetURI, asAmf = false)
            writer.writeUTF(body.responseURI, asAmf = false)

            val bodyWriter = Writer()
            bodyWriter.writeByte(AMF0_AMF3)
            bodyWriter.writeObject(body.data)

            val bodyBytes = bodyWriter.toByteArray()
            writer.writeInt(bodyBytes.size)
            writer.writeBytes(bodyBytes)
        }
    }
}

File: lib/amf/amf3.dart
import 'dart:convert';
import 'dart:typed_data';

/// AMF3 helpers used by web and other pure-Dart targets.
class Amf3 {
  static const int undefinedType = 0;
  static const int nullType = 1;
  static const int falseType = 2;
  static const int trueType = 3;
  static const int integerType = 4;
  static const int doubleType = 5;
  static const int stringType = 6;
  static const int dateType = 8;
  static const int arrayType = 9;
  static const int objectType = 10;
  static const int byteArrayType = 12;
  static const int amf0Amf3 = 17;

  static const int _uint29Mask = 0x1FFFFFFF;
  static const int _int28MaxValue = 0x0FFFFFFF;
  static const int _int28MinValue = -0x0FFFFFFF;

  /// Decode the first body in an AMF envelope.
  static Amf3DecodedBody? decodeFirstBody(Uint8List data) {
    final reader = _Amf3Reader(data);

    reader.readUnsignedShort(); // version
    final headerCount = reader.readUnsignedShort();
    for (var i = 0; i < headerCount; i++) {
      _readHeader(reader);
    }

    final bodyCount = reader.readUnsignedShort();
    if (bodyCount == 0) return null;
    return _readBody(reader);
  }

  static void _readHeader(_Amf3Reader reader) {
    reader.readUtf(); // name
    reader.read(); // mustUnderstand
    for (var i = 0; i < 4; i++) {
      reader.read(); // length
    }
    reader.reset();
    reader.readHeaderObject();
  }

  static Amf3DecodedBody _readBody(_Amf3Reader reader) {
    final target = reader.readUtf();
    final response = reader.readUtf();
    for (var i = 0; i < 4; i++) {
      reader.read(); // length
    }
    reader.reset();
    final data = reader.readObject();
    return Amf3DecodedBody(
      targetURI: target,
      responseURI: response,
      data: data,
    );
  }
}

// ---------------------------------------------------------------------------
// Writer
// ---------------------------------------------------------------------------

class _Amf3Writer {
  final BytesBuilder _builder = BytesBuilder();

  Uint8List toBytes() => _builder.toBytes();

  void writeByte(int v) {
    _builder.addByte(v & 0xFF);
  }

  void writeShort(int v) {
    writeByte((v >> 8) & 0xFF);
    writeByte(v & 0xFF);
  }

  void writeInt(int v) {
    writeByte((v >> 24) & 0xFF);
    writeByte((v >> 16) & 0xFF);
    writeByte((v >> 8) & 0xFF);
    writeByte(v & 0xFF);
  }

  void writeUInt29(int value) {
    var v = value & Amf3._uint29Mask;
    if (v < 0x80) {
      writeByte(v);
    } else if (v < 0x4000) {
      writeByte(((v >> 7) & 0x7F) | 0x80);
      writeByte(v & 0x7F);
    } else if (v < 0x200000) {
      writeByte(((v >> 14) & 0x7F) | 0x80);
      writeByte(((v >> 7) & 0x7F) | 0x80);
      writeByte(v & 0x7F);
    } else if (v < 0x40000000) {
      writeByte(((v >> 22) & 0x7F) | 0x80);
      writeByte(((v >> 15) & 0x7F) | 0x80);
      writeByte(((v >> 8) & 0x7F) | 0x80);
      writeByte(v & 0xFF);
    } else {
      throw ArgumentError('Integer out of range: $value');
    }
  }

  void writeBytes(List<int> bytes) {
    _builder.add(bytes);
  }

  void writeUtf(String value, {required bool asAmf}) {
    final bytes = utf8.encode(value);
    if (asAmf) {
      writeUInt29((bytes.length << 1) | 1);
      writeBytes(bytes);
    } else {
      writeShort(bytes.length);
      writeBytes(bytes);
    }
  }

  void writeStringWithoutType(String value) {
    if (value.isEmpty) {
      writeUInt29(1);
    } else {
      writeUtf(value, asAmf: true);
    }
  }

  void _writeAmfInt(int v) {
    if (v >= Amf3._int28MinValue && v <= Amf3._int28MaxValue) {
      final masked = v & Amf3._uint29Mask;
      writeByte(Amf3.integerType);
      writeUInt29(masked);
    } else {
      writeByte(Amf3.doubleType);
      _writeDouble(v.toDouble());
    }
  }

  void _writeDouble(double value) {
    final bits = Float64List(1)..[0] = value;
    final byteData = bits.buffer.asByteData();
    for (int i = 7; i >= 0; i--) {
      writeByte(byteData.getUint8(i));
    }
  }

  void writeObject(Object? value) {
    if (value == null) {
      writeByte(Amf3.nullType);
    } else if (value is String) {
      writeByte(Amf3.stringType);
      writeStringWithoutType(value);
    } else if (value is bool) {
      writeByte(value ? Amf3.trueType : Amf3.falseType);
    } else if (value is int || value is num) {
      if (value is int) {
        _writeAmfInt(value);
      } else {
        final num n = value as num;
        writeByte(Amf3.doubleType);
        _writeDouble(n.toDouble());
      }
    } else if (value is DateTime) {
      writeByte(Amf3.dateType);
      writeUInt29(1);
      _writeDouble(value.millisecondsSinceEpoch.toDouble());
    } else if (value is List) {
      _writeArray(value);
    } else if (value is Map) {
      _writeMap(value);
    } else if (value is Uint8List || value is List<int>) {
      final bytes =
          value is Uint8List ? value : Uint8List.fromList(value as List<int>);
      writeByte(Amf3.byteArrayType);
      writeUInt29((bytes.length << 1) | 1);
      writeBytes(bytes);
    } else {
      final s = value.toString();
      writeByte(Amf3.stringType);
      writeStringWithoutType(s);
    }
  }

  void _writeArray(List<dynamic> list) {
    writeByte(Amf3.arrayType);
    writeUInt29((list.length << 1) | 1);
    writeUInt29(1);
    for (final item in list) {
      writeObject(item);
    }
  }

  void _writeMap(Map<dynamic, dynamic> map) {
    writeByte(Amf3.objectType);
    writeUInt29(0x0B);
    writeStringWithoutType('');
    map.forEach((k, v) {
      final key = k?.toString() ?? '';
      writeStringWithoutType(key);
      writeObject(v);
    });
    writeStringWithoutType('');
  }
}

// ---------------------------------------------------------------------------
// Reader
// ---------------------------------------------------------------------------

class _Amf3Reader {
  final Uint8List data;
  int pos = 0;

  final List<Object?> objects = [];
  final List<Map<String, Object?>> traits = [];
  final List<String> strings = [];

  _Amf3Reader(this.data);

  int read() => data[pos++] & 0xFF;

  int readUnsignedShort() {
    final c1 = read();
    final c2 = read();
    return (c1 << 8) | c2;
  }

  int readUInt29() {
    int b = read();
    if (b < 0x80) return b;
    int value = (b & 0x7F) << 7;
    b = read();
    if (b < 0x80) return value | b;
    value = (value | (b & 0x7F)) << 7;
    b = read();
    if (b < 0x80) return value | b;
    value = (value | (b & 0x7F)) << 8;
    b = read();
    return value | b;
  }

  Uint8List _readBytes(int length) {
    final bytes = data.sublist(pos, pos + length);
    pos += length;
    return Uint8List.fromList(bytes);
  }

  String readUtf([int? length]) {
    final len = length ?? readUnsignedShort();
    if (len == 0) return '';
    final bytes = _readBytes(len);
    return utf8.decode(bytes);
  }

  void reset() {
    objects.clear();
    traits.clear();
    strings.clear();
  }

  Object? readObject() {
    final type = read();
    return _readObjectValue(type);
  }

  Object? readHeaderObject() {
    var type = read();
    if (type == Amf3.amf0Amf3) {
      type = read();
      return _readObjectValue(type);
    }
    return _readHeaderObjectValue(type);
  }

  String _readString() {
    final ref = readUInt29();
    if ((ref & 1) == 0) {
      final index = ref >> 1;
      return strings[index];
    }
    final len = ref >> 1;
    if (len == 0) return '';
    final str = readUtf(len);
    strings.add(str);
    return str;
  }

  void _rememberObject(Object? v) {
    objects.add(v);
  }

  Object? _getObject(int index) => objects[index];

  Map<String, Object?> _getTraits(int index) => traits[index];

  void _rememberTraits(Map<String, Object?> t) {
    traits.add(t);
  }

  Map<String, Object?> _readTraits(int ref) {
    if ((ref & 3) == 1) {
      return _getTraits(ref >> 2);
    }
    final count = ref >> 4;
    final className = _readString();
    final externalizable = (ref & 4) == 4;
    final dynamic = (ref & 8) == 8;
    final props = <String>[];
    for (var i = 0; i < count; i++) {
      props.add(_readString());
    }
    final traitsMap = <String, Object?>{
      'className': className.isEmpty ? null : className,
      'externalizable': externalizable,
      'dynamic': dynamic,
      'props': props,
    };
    _rememberTraits(traitsMap);
    return traitsMap;
  }

  Object? _readScriptObject() {
    final ref = readUInt29();
    if ((ref & 1) == 0) {
      return _getObject(ref >> 1);
    }

    final traitsMap = _readTraits(ref);
    final externalizable = traitsMap['externalizable'] == true;
    final dynamic = traitsMap['dynamic'] == true;
    final props = (traitsMap['props'] as List).cast<String>();

    final map = <String, Object?>{};
    _rememberObject(map);

    if (externalizable) {
      return map;
    }

    for (final prop in props) {
      final value = readObject();
      map[prop] = value;
    }

    if (dynamic) {
      while (true) {
        final name = _readString();
        if (name.isEmpty) break;
        map[name] = readObject();
      }
    }

    return map;
  }

  Object? _readArray() {
    final ref = readUInt29();
    if ((ref & 1) == 0) {
      return _getObject(ref >> 1);
    }
    final len = ref >> 1;

    Map<String, Object?>? map;
    while (true) {
      final name = _readString();
      if (name.isEmpty) break;
      map ??= <String, Object?>{};
      _rememberObject(map);
      map[name] = readObject();
    }

    if (map == null) {
      final list = List<Object?>.filled(len, null, growable: false);
      _rememberObject(list);
      for (var i = 0; i < len; i++) {
        list[i] = readObject();
      }
      return list;
    } else {
      for (var i = 0; i < len; i++) {
        map[i.toString()] = readObject();
      }
      return map;
    }
  }

  double _readDouble() {
    final b0 = read();
    final b1 = read();
    final b2 = read();
    final b3 = read();
    final b4 = read();
    final b5 = read();
    final b6 = read();
    final b7 = read();
    final bytes = Uint8List.fromList([b7, b6, b5, b4, b3, b2, b1, b0]);
    final bd = ByteData.view(bytes.buffer);
    return bd.getFloat64(0, Endian.big);
  }

  Object? _readDate() {
    final ref = readUInt29();
    if ((ref & 1) == 0) {
      return _getObject(ref >> 1);
    }
    final time = _readDouble().toInt();
    final date = DateTime.fromMillisecondsSinceEpoch(time, isUtc: true);
    _rememberObject(date);
    return date;
  }

  Map<String, Object?> _readMap() {
    final ref = readUInt29();
    if ((ref & 1) == 0) {
      return _getObject(ref >> 1) as Map<String, Object?>;
    }
    final length = ref >> 1;
    final map = <String, Object?>{};
    if (length > 0) {
      _rememberObject(map);
      while (true) {
        final name = readObject();
        if (name is! String) break;
        map[name] = readObject();
      }
    }
    return map;
  }

  Uint8List _readByteArray() {
    final ref = readUInt29();
    if ((ref & 1) == 0) {
      return _getObject(ref >> 1) as Uint8List;
    }
    final len = ref >> 1;
    final bytes = _readBytes(len);
    _rememberObject(bytes);
    return bytes;
  }

  Object? _readHeaderObjectValue(int type) {
    switch (type) {
      case 2:
        return readUtf();
      default:
        throw StateError('Unknown AMF0 header type: $type');
    }
  }

  Object? _readObjectValue(int type) {
    switch (type) {
      case Amf3.stringType:
        return _readString();
      case Amf3.objectType:
        return _readScriptObject();
      case Amf3.arrayType:
        return _readArray();
      case Amf3.falseType:
        return false;
      case Amf3.trueType:
        return true;
      case Amf3.integerType:
        final temp = readUInt29();
        return (temp << 3) >> 3;
      case Amf3.doubleType:
        return _readDouble();
      case Amf3.undefinedType:
      case Amf3.nullType:
        return null;
      case Amf3.dateType:
        return _readDate();
      case Amf3.byteArrayType:
        return _readByteArray();
      case Amf3.amf0Amf3:
        return readObject();
      default:
        throw StateError('Unknown AMF3 type: $type');
    }
  }
}

// ---------------------------------------------------------------------------
// ActionMessage and serializer
// ---------------------------------------------------------------------------

class Amf3MessageBody {
  final String targetURI;
  final String responseURI;
  final Object? data;

  const Amf3MessageBody({
    required this.targetURI,
    required this.responseURI,
    required this.data,
  });
}

class Amf3ActionMessage {
  final int version;
  final List<Amf3MessageBody> bodies;

  const Amf3ActionMessage({
    this.version = 3,
    required this.bodies,
  });
}

class Amf3Serializer {
  Uint8List writeMessage(Amf3ActionMessage message) {
    final writer = _Amf3Writer();
    writer.writeShort(message.version);
    writer.writeShort(0); // header count
    writer.writeShort(message.bodies.length);

    for (final body in message.bodies) {
      _writeBody(writer, body);
    }

    return writer.toBytes();
  }

  void _writeBody(_Amf3Writer writer, Amf3MessageBody body) {
    writer.writeUtf(body.targetURI, asAmf: false);
    writer.writeUtf(body.responseURI, asAmf: false);

    final bodyWriter = _Amf3Writer();
    bodyWriter.writeByte(Amf3.amf0Amf3);
    bodyWriter.writeObject(body.data);

    final bodyBytes = bodyWriter.toBytes();
    writer.writeInt(bodyBytes.length);
    writer.writeBytes(bodyBytes);
  }
}

class Amf3DecodedBody {
  final String targetURI;
  final String responseURI;
  final Object? data;

  const Amf3DecodedBody({
    required this.targetURI,
    required this.responseURI,
    required this.data,
  });
}
